src/main/java/epicsquid/roots/util/RitualUtil.java

```diff
@@ -12,30 +13,45 @@
 import vazkii.patchouli.common.book.Book;
 import vazkii.patchouli.common.book.BookRegistry;
 
+import java.util.Arrays;
+import java.util.List;
 import java.util.Random;
 
 public class RitualUtil {
 
   private static Random rand = new Random();
 
-    public static BlockPos getRandomPosRadialXZ(BlockPos centerPos, int xRadius, int zRadius)
-    {
-        BlockPos pos = centerPos.add(-xRadius, 0, -zRadius);
+  public static BlockPos getRandomPosRadialXZ(BlockPos centerPos, int xRadius, int zRadius)
+  {
+      BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(centerPos.getX() -xRadius, centerPos.getY(), centerPos.getZ() -zRadius);
 
-        pos = pos.add(rand.nextInt(xRadius*2), 0, rand.nextInt(zRadius*2));
+      return pos.add(rand.nextInt(xRadius * 2), 0, rand.nextInt(zRadius * 2));
+  }
 
-        //System.out.println("Pos: " +  pos.getX() +  " | " + pos.getY() + " | " + pos.getZ());
-        return pos;
-    }
+  public static BlockPos getRandomPosRadialXYZ(BlockPos centerPos, int xRadius, int yRadius, int zRadius)
+  {
+      BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(centerPos.getX() -xRadius, centerPos.getY() -yRadius, centerPos.getZ() -zRadius);
 
-    public static BlockPos getRandomPosRadialXYZ(BlockPos centerPos, int xRadius, int yRadius, int zRadius)
-    {
-        BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(centerPos.getX() -xRadius, centerPos.getY() -yRadius, centerPos.getZ() -zRadius);
+      BlockPos pos2 = pos.add(rand.nextInt(xRadius * 2), rand.nextInt(yRadius * 2), rand.nextInt(zRadius * 2));
 
-        BlockPos pos2 = pos.add(rand.nextInt(xRadius * 2), rand.nextInt(yRadius * 2), rand.nextInt(zRadius * 2));
-        //System.out.println("Pos: " +  pos.getX() +  " | " + pos.getY() + " | " + pos.getZ());
-        return pos2;
-    }
+      //Debug Print
+      //System.out.println("Pos: " +  pos.getX() +  " | " + pos.getY() + " | " + pos.getZ());
+      return pos2;
+  }
+
+  public static BlockPos getRandomPosRadialXYZ(World world, BlockPos centerPos, int xRadius, int yRadius, int zRadius, Block... whitelistedBlocks)
+  {
+    BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(centerPos.getX() -xRadius, centerPos.getY() -yRadius, centerPos.getZ() -zRadius);
```

Comment:
Mutable is mostly when you're iterating over a pre-defined area and you want to save on memory usage by reusing the same BlockPos *object* instance but changing its X/Y/Z values. You also need to make sure that you use "toImmutable" or whatever it's called now before passing it back unless you're explicitly returning a Mutable Block Pos.

In this case, `pos.add` actually changes `pos` (I think...) so you don't even need to declare the second object. I don't know that it really saves much though, because all you're doing is reducing the instantiation of a second BlockPos object.

`BlockPos.getAllInBoxMutable` is a really good example of its usage. It creates one object and just increments/decrements the x, y and z values as necessary instead of creating an object for each time `.next()` is called. If you're doing a large space this can save on object creation costs.

</random waffle>

Reply:
oki, I'll check it out later